### PR TITLE
TBB stream executor

### PR DESCRIPTION
**Pull request by apankratovantonp**
* Implemented CPU executor with same functionality as Stream CPU Executor but it used `TBB` threads only
* It should improve IE interaction with other libraries. Especially with that uses `TBB` as threading back end (`OpenCV`)

### CODE REVIEW

**Comment by dmitry-gorokhov** [comment]
@apankratovantonp Do we have strong requirement to put this into 2020.4 release or we can merge it right after release branch creation?

**Comment by apankratovantonp** [comment]
Can be postponed.

**Comment by ilya-lavrenov** [comment]
@apankratovantonp looks like this branch does not contain TBB stream executor, but the same content as https://github.com/openvinotoolkit/openvino/pull/528

File: inference-engine/src/inference_engine/threading/ie_tbb_streams_executor.cpp [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+#include "ie_system_conf.h"
+#include "details/ie_exception.hpp"
+
+#if ((IE_THREAD == IE_THREAD_TBB) || (IE_THREAD == IE_THREAD_TBB_AUTO))
+#include <tbb/task_scheduler_observer.h>
+#include <tbb/task_arena.h>
+#include <tbb/task_group.h>
+#include <tbb/concurrent_queue.h>
+#include <tbb/enumerable_thread_specific.h>
+// #include <tbb/task_scheduler_init.h>
```
Comment (ilyachur) [inline inference-engine/src/inference_engine/threading/ie_tbb_streams_executor.cpp:26]: nit: redundant comment

File: inference-engine/src/plugin_api/threading/ie_tbb_streams_executor.hpp [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+/**
+* @class TBBStreamsExecutor
+* @brief CPU Streams executor implementation. Use TBB thread pool to run tasks
+*/
+class INFERENCE_ENGINE_API_CLASS(TBBStreamsExecutor) : public IStreamsExecutor {
+public:
+    using Ptr = std::shared_ptr<TBBStreamsExecutor>;
+    explicit TBBStreamsExecutor(const Config& config = {});
+    ~TBBStreamsExecutor() override;
+    void run(Task task) override;
```
Comment (ilyachur) [inline inference-engine/src/plugin_api/threading/ie_tbb_streams_executor.hpp:23]: Can we change `Task` to `const Task&`?

File: inference-engine/src/mkldnn_plugin/mkldnn_exec_network.cpp [hunk 3]
```diff
@@ -249,12 +300,12 @@ InferenceEngine::Parameter MKLDNNExecNetwork::GetMetric(const std::string &name)
         IE_SET_METRIC_RETURN(SUPPORTED_METRICS, metrics);
     } else if (name == METRIC_KEY(SUPPORTED_CONFIG_KEYS)) {
         std::vector<std::string> configKeys;
-        for (auto && key : _graphs.begin()->get()->getProperty()._config) {
+        for (auto && key : const_cast<MKLDNNExecNetwork*>(this)->GetGraph()._graph.getProperty()._config) {
```
Comment (ilyachur) [inline inference-engine/src/mkldnn_plugin/mkldnn_exec_network.cpp:303]: Can we avoid `const_cast`?

File: inference-engine/src/plugin_api/threading/ie_tbb_streams_executor.hpp (hunk 2 above)
Comment (apankratovantonp) [reply inference-engine/src/plugin_api/threading/ie_tbb_streams_executor.hpp:23]: No. Intended to use with move semantic.

File: inference-engine/src/mkldnn_plugin/mkldnn_exec_network.cpp (hunk 3 above)
Comment (apankratovantonp) [reply inference-engine/src/mkldnn_plugin/mkldnn_exec_network.cpp:303]: Can do some members `mutable`

**Comment by ilyachur** [comment]
@dmitry-gorokhov Please have a look.

**Comment by myshevts** [comment]
@apankratovantonp that's too much changes to review without having some general feature into.
THere are not enough comments in the code, so I have to reverse engineer the logic, whihc is hard and error prone.
I suggest having a short meeting with live review

**Comment by ilya-lavrenov** [comment]
@apankratovantonp is it ready for review? can we have a discussion Maxim is talking about?

**Comment by apankratovantonp** [comment]
> @apankratovantonp is it ready for review? can we have a discussion Maxim is talking about?

@ilya-lavrenov  Yes it is ready. We already have some discussion with @myshevts. He was going to review the code deeper and has another one discussion.

**Review by myshevts — COMMENTED** [review]
I'm generally fine now, yet would kindly ask to wait for the https://github.com/openvinotoolkit/openvino/pull/3199 (ADL therading) to be merged first, to see if/how we can factorize the code and avoid duplication (e.g. on the selected numaNodeID, streamsID, and <new>core type logic) - as we discussed this can be a common base class.

**Review by myshevts — APPROVED** [review]

**Comment by apankratovantonp** [comment]
@dmitry-gorokhov Please have a look

**Comment by apankratovantonp** [comment]
@dmitry-gorokhov ping

**Comment by myshevts** [comment]
as I mentioned in the [previous comment](https://github.com/openvinotoolkit/openvino/pull/557#pullrequestreview-569334483) 
this PR may introduce issues with hybrid support that is being implemented
.
we are currently fighting to meet a feature freeze for the hybrid support with @ivankochin , and I would kindly ask to postpone the merge of this PR, since it would require additional work from our side, and the time is too tight for us.

**Comment by dmitry-gorokhov** [comment]
@myshevts @apankratovantonp  As far as I know all hybrid-aware related functionality was merged recently. If so can we proceed with this PR?

**Comment by myshevts** [comment]
yes, we should continue with this PR again

**Comment by apankratovantonp** [comment]
@dmitry-gorokhov Can be merged

**Review by myshevts — APPROVED** [review]

File: inference-engine/src/mkldnn_plugin/mkldnn_exec_network.cpp [hunk 4]
```diff
@@ -35,6 +38,14 @@ MKLDNNExecNetwork::CreateInferRequestImpl(InferenceEngine::InputsDataMap network
     return std::make_shared<MKLDNNInferRequest>(networkInputs, networkOutputs, std::static_pointer_cast<MKLDNNExecNetwork>(shared_from_this()));
 }
 
+struct ImmediateSerialExecutor : public ITaskExecutor {
+    void run(InferenceEngine::Task task) override {
+        std::unique_lock<std::mutex> l{_mutex};
```
[the remark this answers is not in the document]
Comment (apankratovantonp) [reply inference-engine/src/mkldnn_plugin/mkldnn_exec_network.cpp]: done

**Comment by apankratovantonp** [comment]
@dmitry-gorokhov Can be merged

**Review by dmitry-gorokhov — APPROVED** [review]

**Pull request merged**